### ISSUE TITLE
[LightGBM] [Fatal] Multiclass objective and metrics don't match Error in data$construct() : Multiclass objective and metrics don't match

### ISSUE THREAD

**Issue by SHG369** [microsoft/LightGBM#6269]
## Description
I'm going to work on my  data with an Light Gradient Boosting Machine.
My data is like this.
> head(gbmdat)[1:4,1:4]
           Group    CLEC4D   MCEMP1    C3AR1
G2M1     0 10.357580 8.105276 6.842050
G2M2     1  8.307793 7.534084 8.452518
G2M3     2 10.215788 8.745289 9.006207
G2M4     0  9.174830 8.599312 7.778305

column “Group” was choose to the label:

dtrain <- lgb.Dataset(data=as.matrix(gbmdat[,-1]), 
                      label =gbmdat$Group)
params like this :
params <- list(
  boosting_type ="gbdt",
  objective = 'multiclass',
  metric = "multi_logloss", 
  nthread = 4,
  learning_rate = 0.05,
  max_depth = 5,
  num_leaves = 40,
  feature_fraction = 0.1,  
  bagging_fraction = 0.1,
  bagging_freq = 1,
  num_class = 3 
)

Then the error occurred when the code runs:
> lgb_model <- lgb.train( params = params,
+                         data = dtrain,
+                         nrounds = 5000,
+                         verbose = -1)
[LightGBM] [Fatal] Multiclass objective and metrics don't match
Error in data$construct() : Multiclass objective and metrics don't match

So why there is an error ?
could you please help me ?
thanks.

**Comment by jameslamb** [comment]
Thanks for using LightGBM.

---

I've reformatted your post to make the distinction between your own words, code, and output from code clearer. If you're not sure how to do that, please review https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax.

---

In the future, we would also appreciate if you could provide a minimal, reproducible example.

* "minimal" = "the least and fastest code necessary to reproduce the behavior"
  - *is it really necessary to produce 5000 trees each with 40 leaves to observe this error?*
  - *does this error show up with all of LightGBMM's default arguments?*
* "reproducible" meaning "with all the details needed for someone else to run the code and see the same behavior you're seeing".
   - *what version of R? version of `{lightgbm}`?*
   - *how did you install `{lightgbm}`?*
   - *what operating system are you using?*

You can learn more about that at https://stackoverflow.com/help/minimal-reproducible-example.

For example, using `{lightgbm}` 4.2.0 (the latest version on CRAN), R 4.2.3, and macOS, and the following code that matches yours closely (with guesses for the other parts that were omitted), I do not get an error.

```r
library(lightgbm)

gbmdat <- data.frame(
    Group = sample(c(0L, 1L, 2L), size = 1000, replace = TRUE)
    , CLECL4D = rnorm(n = 1000)
    , MCEMP1 = rnorm(n = 1000)
    , C3AR1 = rnorm(n = 1000)
)

dtrain <- lightgbm::lgb.Dataset(
    data = as.matrix(gbmdat[,-1])
    , label = gbmdat$Group
)

params <- list(
    boosting_type ="gbdt"
    , objective = "multiclass"
    , metric = "multi_logloss"
    , nthread = 4
    , learning_rate = 0.05
    , max_depth = 5
    , num_leaves = 5
    , feature_fraction = 0.1
    , bagging_fraction = 0.1
    , bagging_freq = 1
    , num_class = 3
)

lgb_model <- lgb.train(
    params = params
    , data = dtrain
    , nrounds = 10
    , verbose = -1
)

summary(lgb_model)

# LightGBM Model (10 trees)
# Objective: multiclass (3 classes)
# Fitted to dataset with 3 columns
```

We'd be happy to help you, but please provide the details I've asked for above.

**Comment by SHG369** [comment]
> version
               _                                
platform       x86_64-w64-mingw32               
version.string R version 4.3.1 (2023-06-16 ucrt)
lightgbm version 4.2.0 



Very thanks for help me. Actually, the code provided in your reply didn't work on my computer. A same error occurred again.

**Comment by SHG369** [comment]
> Thanks for using LightGBM.
> 
> I've reformatted your post to make the distinction between your own words, code, and output from code clearer. If you're not sure how to do that, please review https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax.
> 
> In the future, we would also appreciate if you could provide a minimal, reproducible example.
> 
> * "minimal" = "the least and fastest code necessary to reproduce the behavior"
>   
>   * _is it really necessary to produce 5000 trees each with 40 leaves to observe this error?_
>   * _does this error show up with all of LightGBMM's default arguments?_
> * "reproducible" meaning "with all the details needed for someone else to run the code and see the same behavior you're seeing".
>   
>   * _what version of R? version of `{lightgbm}`?_
>   * _how did you install `{lightgbm}`?_
>   * _what operating system are you using?_
> 
> You can learn more about that at https://stackoverflow.com/help/minimal-reproducible-example.
> 
> For example, using `{lightgbm}` 4.2.0 (the latest version on CRAN), R 4.2.3, and macOS, and the following code that matches yours closely (with guesses for the other parts that were omitted), I do not get an error.
> 
> ```r
> library(lightgbm)
> 
> gbmdat <- data.frame(
>     Group = sample(c(0L, 1L, 2L), size = 1000, replace = TRUE)
>     , CLECL4D = rnorm(n = 1000)
>     , MCEMP1 = rnorm(n = 1000)
>     , C3AR1 = rnorm(n = 1000)
> )
> 
> dtrain <- lightgbm::lgb.Dataset(
>     data = as.matrix(gbmdat[,-1])
>     , label = gbmdat$Group
> )
> 
> params <- list(
>     boosting_type ="gbdt"
>     , objective = "multiclass"
>     , metric = "multi_logloss"
>     , nthread = 4
>     , learning_rate = 0.05
>     , max_depth = 5
>     , num_leaves = 5
>     , feature_fraction = 0.1
>     , bagging_fraction = 0.1
>     , bagging_freq = 1
>     , num_class = 3
> )
> 
> lgb_model <- lgb.train(
>     params = params
>     , data = dtrain
>     , nrounds = 10
>     , verbose = -1
> )
> 
> summary(lgb_model)
> 
> # LightGBM Model (10 trees)
> # Objective: multiclass (3 classes)
> # Fitted to dataset with 3 columns
> ```
> 
> We'd be happy to help you, but please provide the details I've asked for above.

version
_
platform x86_64-w64-mingw32
version.string R version 4.3.1 (2023-06-16 ucrt)
lightgbm version 4.2.0

Very thanks for help me. Actually, the code provided in your reply didn't work on my computer. A same error occurred again.

**Comment by jameslamb** [comment]
Here is where that error comes from:

https://github.com/microsoft/LightGBM/blob/2e3543c04c4b8e475966721ad8570544edd5b470/src/io/config.cpp#L319-L322

We do test the R package against R 4.3.1 on Windows:

https://github.com/microsoft/LightGBM/blob/2e3543c04c4b8e475966721ad8570544edd5b470/.ci/test_r_package_windows.ps1#L85

We need some more information:

* how did you install `{lightgbm}`? What command(s) did you run?
* Can you please share the output of running `sessionInfo()` in R?

**Comment by SHG369** [comment]
> Here is where that error comes from:
> 
> https://github.com/microsoft/LightGBM/blob/2e3543c04c4b8e475966721ad8570544edd5b470/src/io/config.cpp#L319-L322
> 
> We do test the R package against R 4.3.1 on Windows:
> 
> https://github.com/microsoft/LightGBM/blob/2e3543c04c4b8e475966721ad8570544edd5b470/.ci/test_r_package_windows.ps1#L85
> 
> We need some more information:
> 
> * how did you install `{lightgbm}`? What command(s) did you run?
> * Can you please share the output of running `sessionInfo()` in R?

Simply code install.packages("lightgbm") to install.
> sessionInfo()
R version 4.3.1 (2023-06-16 ucrt)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 18363)

Matrix products: default


locale:
[1] LC_COLLATE=Chinese (Simplified)_China.utf8  LC_CTYPE=Chinese (Simplified)_China.utf8    LC_MONETARY=Chinese (Simplified)_China.utf8
[4] LC_NUMERIC=C                                LC_TIME=Chinese (Simplified)_China.utf8    

time zone: Asia/Shanghai
tzcode source: internal

attached base packages:
[1] tcltk     stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] xgboost_1.7.6.1       lightgbm_4.2.0        rpart_4.1.21          bestglm_0.37.3        leaps_3.1             CoxBoost_1.5         
 [7] prodlim_2023.08.28    survival_3.5-7        WGCNA_1.72-5          fastcluster_1.2.3     dynamicTreeCut_1.63-1 preprocessCore_1.64.0
[13] devtools_2.4.5        usethis_2.2.2         randomForest_4.7-1.1  glmnet_4.1-8          Matrix_1.6-1.1        lubridate_1.9.3      
[19] forcats_1.0.0         purrr_1.0.2           readr_2.1.4           tidyr_1.3.0           tibble_3.2.1          ggplot2_3.4.4        
[25] tidyverse_2.0.0       stringr_1.5.0         Mfuzz_2.62.0          DynDoc_1.80.0         widgetTools_1.80.0    e1071_1.7-14         
[31] Biobase_2.62.0        BiocGenerics_0.48.0   dplyr_1.1.3           limma_3.58.0          TCseq_1.26.0         

loaded via a namespace (and not attached):
  [1] splines_4.3.1               later_1.3.1                 bitops_1.0-7                hardhat_1.3.0              
  [5] pROC_1.18.5                 graph_1.80.0                XML_3.99-0.14               lifecycle_1.0.3            
  [9] rstatix_0.7.2               edgeR_4.0.2                 doParallel_1.0.17           rprojroot_2.0.3            
 [13] MASS_7.3-60                 globals_0.16.2              processx_3.8.2              lattice_0.21-9             
 [17] backports_1.4.1             magrittr_2.0.3              rmarkdown_2.25              Hmisc_5.1-1                
 [21] remotes_2.4.2.1             httpuv_1.6.12               sessioninfo_1.2.2           pkgbuild_1.4.2             
 [25] DBI_1.1.3                   RColorBrewer_1.1-3          abind_1.4-5                 pkgload_1.3.3              
 [29] zlibbioc_1.48.0             GenomicRanges_1.54.0        RCurl_1.98-1.12             nnet_7.3-19                
 [33] ipred_0.9-14                lava_1.7.3                  circlize_0.4.15             GenomeInfoDbData_1.2.11    
 [37] IRanges_2.36.0              S4Vectors_0.40.1            listenv_0.9.0               parallelly_1.36.0          
 [41] annotate_1.80.0             codetools_0.2-19            DelayedArray_0.28.0         tidyselect_1.2.0           
 [45] shape_1.4.6                 farver_2.1.1                matrixStats_1.0.0           stats4_4.3.1               
 [49] base64enc_0.1-3             jsonlite_1.8.7              GenomicAlignments_1.38.0    caret_6.0-94               
 [53] GetoptLong_1.0.5            ellipsis_0.3.2              Formula_1.2-5               iterators_1.0.14           
 [57] foreach_1.5.2               tools_4.3.1                 Rcpp_1.0.11                 glue_1.6.2                 
 [61] gridExtra_2.3               SparseArray_1.2.0           xfun_0.40                   MatrixGenerics_1.14.0      
 [65] GenomeInfoDb_1.38.0         withr_2.5.1                 BiocManager_1.30.22         fastmap_1.1.1              
 [69] fansi_1.0.5                 callr_3.7.3                 digest_0.6.33               timechange_0.2.0           
 [73] R6_2.5.1                    mime_0.12                   colorspace_2.1-0            GO.db_3.18.0               
 [77] RSQLite_2.3.2               utf8_1.2.4                  generics_0.1.3              recipes_1.0.9              
 [81] pls_2.8-3                   data.table_1.14.8           class_7.3-22                prettyunits_1.2.0          
 [85] httr_1.4.7                  htmlwidgets_1.6.2           S4Arrays_1.2.0              ModelMetrics_1.2.2.2       
 [89] pkgconfig_2.0.3             gtable_0.3.4                timeDate_4032.109           blob_1.2.4                 
 [93] impute_1.76.0               ComplexHeatmap_2.18.0       XVector_0.42.0              htmltools_0.5.6.1          
 [97] carData_3.0-5               profvis_0.3.8               GSEABase_1.64.0             clue_0.3-65                
[101] scales_1.2.1                png_0.1-8                   tkWidgets_1.80.0            gower_1.0.1                
[105] knitr_1.44                  rstudioapi_0.15.0           tzdb_0.4.0                  reshape2_1.4.4             
[109] rjson_0.2.21                nlme_3.1-163                checkmate_2.3.1             curl_5.1.0                 
[113] proxy_0.4-27                cachem_1.0.8                GlobalOptions_0.1.2         parallel_4.3.1             
[117] miniUI_0.1.1.1              foreign_0.8-85              AnnotationDbi_1.64.0        desc_1.4.2                 
[121] pillar_1.9.0                grid_4.3.1                  vctrs_0.6.4                 urlchecker_1.0.1           
[125] promises_1.2.1              ggpubr_0.6.0                car_3.1-2                   xtable_1.8-4               
[129] cluster_2.1.4               htmlTable_2.4.2             evaluate_0.22               cli_3.6.1                  
[133] locfit_1.5-9.8              compiler_4.3.1              Rsamtools_2.18.0            rlang_1.1.1                
[137] crayon_1.5.2                future.apply_1.11.0         ggsignif_0.6.4              grpreg_3.4.0               
[141] labeling_0.4.3              ps_1.7.5                    plyr_1.8.9                  fs_1.6.3                   
[145] stringi_1.7.12              BiocParallel_1.36.0         munsell_0.5.0               Biostrings_2.70.1          
[149] hms_1.1.3                   future_1.33.0               bit64_4.0.5                 KEGGREST_1.42.0            
[153] statmod_1.5.0               shiny_1.7.5.1               SummarizedExperiment_1.32.0 broom_1.0.5                
[157] memoise_2.0.1               bit_4.0.5

**Comment by Elaine-88** [comment]
Have you handled this problem? I meet the same error report

**Comment by jameslamb** [comment]
@Elaine-88 can you provide the same things I asked for in https://github.com/microsoft/LightGBM/issues/6269#issuecomment-1888492138 and https://github.com/microsoft/LightGBM/issues/6269#issuecomment-1888384660?

Do you experience this error with the example code I provided there? If not, can you provide some **minimal** code that reproduces it?

**Comment by Elaine-88** [comment]
> [@Elaine-88](https://github.com/Elaine-88) can you provide the same things I asked for in [#6269 (comment)](https://github.com/microsoft/LightGBM/issues/6269#issuecomment-1888492138) and [#6269 (comment)](https://github.com/microsoft/LightGBM/issues/6269#issuecomment-1888384660)?
> 
> Do you experience this error with the example code I provided there? If not, can you provide some **minimal** code that reproduces it?

Thank you, I have handled it.

**Comment by jameslamb** [comment]
Oh great, glad it's working for you!

What was the problem and what was the solution?
